### PR TITLE
feat(core): port the session domain from KroApple

### LINKED ISSUE #8: Session domain: configs & presets, anchored fragments, running-session state, recommendation, points formulas
Part of #1 (Kro Web parity epic).

**What.** `packages/core/src/domain/session/**`: `SessionTimerMode` (countdown/stopwatch), `SessionConfig` + canonical presets (Quick Focus 5m · Pomodoro 25+5 · Focus 50+10 · Momentum 75+15 · Headspace 3h+60 · Open Space stopwatch), `SessionFragment`, `PersistedRunningSession` (+phases running/paused/break/concluded) with **anchored time accounting** — elapsed = Σ (end ?? now) − start; display ticks never persist; `remainingDuration` clamps at 0 — `SessionLaunchRecommendation` (preferred → empirical mean of ≥3 qualifying performances within min/max bounds → stopwatch → configured default; source enum incl. `empirical(sampleCount)`), the **30%-of-target abort threshold**, and both **reward formulas**: sliding scale (finished+completed 100% / finished-only 30% / early-finish proportional / early-not-completed 0% / quick-complete 100% / aborted no record) and legacy (est. minutes × priority multiplier 1.5×/1.25×/1× × resolution factor).

**Why.** These exact semantics are the product's core loop; the existing `useSession`/`useSessionTimer` code approximates a subset and will be rebuilt on this (child #21).

**Scenario:** `react-uzf-v1` · Stack Matrix Scenario 5 (Web Only)
**UZF pattern & constraints:** pure domain, no timers/side effects here (`RC-24`); Producers wire clocks later.
Canon: `KroCore/Model/Session/*`, `docs/Features/Session.md`, `docs/Features/Performances.md`, `SessionSetupFeature.swift` (threshold). Canon: `zheref/KroApple@2c1ee45` (re-fetch `origin/main` before building; note divergence in the PR).

**File lane (exclusive):** `packages/core/src/domain/session/**`.

**Blocked by:** #7 · **Blocks:** #21.

**Acceptance criteria**
1. Pause/resume/kill-restore scenarios recompute wall-clock-correct elapsed/remaining purely from fragments + now.
2. Recommendation priority matches canon for all four sources, including the "learned until explicitly made Preferred" rule.
3. Both formulas reproduce the canon tables exactly (worked examples as test fixtures).

**Test expectations:** ≥3 tests per formula branch and per recommendation source; property test: Σ fragments never drifts under arbitrary pause/resume sequences.
**Analytics/Distribution:** n/a.
**Out of scope:** UI, persistence of the anchor file (#10), the feature slice (#21).

### PR DESCRIPTION
**Pull request by zheref**
> 🟥 **Ichigo · Substitute** — the local plane, entire · *local, on your creds · I open PRs for **you** to merge (I never merge `main`, never review my own work)*
> Acting as **Shinigami** (product code). Authored locally in a Claude Code session on behalf of @zheref — no CI run.

Closes #8 · Part of #1

# What this changes for you

Kro Web now knows how a focus session actually works — the same way the iPhone app does, to the number.

Start a 25-minute session, pause it, close the tab, come back seven minutes later: the timer shows the *right* remaining time, not the time it had when you left. That is because nothing in here counts ticks. Every figure on screen is recomputed from a list of timestamped fragments against the current moment, which is what makes the session survive a kill, a throttled background tab, or a laptop lid closing for an hour.

It also knows what a session was *worth*. Finish the timer and mark the task done — full points. Finish the timer but leave the task open — 30%. Stop at 15 of 25 minutes with the task done — 60%, proportionally. Stop at 4 minutes — that was an aborted attempt, worth nothing, and it deliberately does not teach Kro that this task "takes 4 minutes". And when you open a session on an endeavor, it opens the way KroApple would: your preferred duration if you set one, otherwise a duration learned from your last few sessions, otherwise an open-ended stopwatch.

None of this is wired to a screen yet — that is #21 (session feature logic) and #22 (session UI). This PR is the rulebook they will both read from, and it is the reason those two can be built without re-deriving any of it.

## What changed

Everything lands in `packages/core/src/domain/session/**` (this issue's exclusive lane), plus the two barrel entries. Pure values and pure functions only — no timers, no storage, no clock (`RC-24`); every time-relative function takes `now` as a parameter, the precedent #7 set.

| File | What it carries |
|---|---|
| `FocusTimerMode.ts` | countdown / stopwatch, with raw-value narrowing for a stored anchor |
| `FocusSessionConfig.ts` | the config value + the six canonical presets (Quick Focus 5m · Pomodoro 25+5 · Focus 50+10 · Momentum 75+15 · Headspace 3h+60m · Open Space stopwatch) |
| `FocusSessionFragment.ts` | one continuous focus period, its span against `now`, and the one sanctioned conversion to `PerformFragment` |
| `PersistedRunningSession.ts` | the anchor — endeavor identity, target, mode, fragments, phase — plus `elapsed`/`remaining` and the structural-consistency predicate |
| `SessionTransitions.ts` | the pure phase moves: `pauseSessionAt` · `resumeSessionAt` · `concludeSessionAt` · `startBreakAt` · `closeSessionAt` |
| `SessionThreshold.ts` | the 30%-of-target recording threshold and the finish-early outcome |
| `SessionLaunchRecommendation.ts` | the four-rung launch priority + `empiricalDuration` / `empiricalDurationPerformances` (which #7 explicitly deferred here) |
| `PointsFormula.ts` · `RewardCalculator.ts` | both reward formulas as pure functions, behind one `awardRewardPoints` entry point |
| `__mocks__/` | five fixture spreads (`RC-13`), ≥7 variants each, exported via `@kro/core/mocks` |

**Tests: 245 in this lane** (716 in `packages/core` overall), including the binding ones from the issue — ≥3 per formula branch and per recommendation source, threshold boundary cases at exactly 30% and just under, kill-restore scenarios through a real JSON round trip, and the drift property test below.

**Coverage on touched files: 99.61%.** Every file is 100% except `RewardCalculator.ts` at 96.34% — the three uncovered lines are its `assertNever` default arms, which are unreachable by construction (`RC-9`). That is the only exemption claimed.

## How to verify

```bash
git fetch origin && git checkout ichigo/8-session-domain
make setup

# The three gates, all green on this branch:
pnpm --filter @kro/core lint     # platform-free check
make typecheck                   # tsc --noEmit across every member
make test                        # every member's Vitest suite

# Just this lane (245 tests):
pnpm --filter @kro/core exec vitest run src/domain/session
```

**Coverage.** `@vitest/coverage-v8` is a dependency of `apps/web`, not of `packages/core`, and this PR adds no dependency (the lockfile is untouched) — so the number above was measured by temporarily linking the provider into `packages/core/node_modules`, then removing it. To reproduce:

```bash
mkdir -p packages/core/node_modules/@vitest
ln -sfn "$PWD/apps/web/node_modules/@vitest/coverage-v8" packages/core/node_modules/@vitest/coverage-v8
pnpm --filter @kro/core exec vitest run --coverage --coverage.provider=v8 \
  --coverage.include='src/domain/session/**' --coverage.all=true
rm -rf packages/core/node_modules/@vitest
```

### Sliding scale — every row of `docs/Features/Performances.md`'s summary table

Base points are canon's flat 30 (`baseRewardFor`); target is the doc's 25-minute example (1500 s).

| Canon doc row | Resolution | Elapsed | Canon says | This PR | Test |
|---|---|---|---|---|---|
| Timer finished, task **not** completed | `complete` | 1500 (100%) | 30% of base | **9** | `RewardCalculator.test.ts` |
| Timer finished, task completed | `finished` | 1500 (100%) | 100% of base | **30** | ″ |
| Finished early, task completed *(doc's worked example)* | `finished` | 900 (60%) | 60% of base | **18** | ″ |
| Finished early, task **not** completed | `complete` | 900 (60%) | 0% | **0** | ″ |
| Quick complete, no session | `finished` | 0, target 0 | 100% of base | **30** | ″ |
| Aborted | `aborted` | any | 0 / not awarded | **0** | ″ |

Two boundary behaviours worth eyeballing, both canon's and both pinned by tests: the `finished` arm **caps at 100%** so overrunning never pays more (elapsed 5400 → still 30), and it **truncates** rather than rounds — 1300 of 2000 s is exactly 65%, 65% of 30 is exactly 19.5, and canon's `Int(Double)` pays **19**, not 20.

### Legacy formula — `estimated_minutes × 1 pt/min × priority_multiplier × resolution_factor`

Estimate is 25 minutes; `now` is 2026-01-15 09:00 local.

| Due date | Multiplier | Resolution | Raw | This PR | 
|---|---|---|---|---|
| none | 1× | `finished` | 25 | **25** |
| none | 1× | `complete` | 7.5 | **8** |
| none | 1× | `aborted` | 0 | **0** |
| 08:00 (overdue) | 1.5× | `finished` | 37.5 | **38** |
| 08:00 (overdue) | 1.5× | `complete` | 11.25 | **11** |
| 10:00 (in 1 h) | 1.25× | `finished` | 31.25 | **31** |
| 11:00 (in exactly 2 h) | 1.25× | `finished` | 31.25 | **31** |
| 11:00:01 (2 h + 1 s) | 1× | `finished` | 25 | **25** |

Note the deliberate asymmetry against the sliding scale: legacy **rounds** (canon's `.rounded()`), so 7.5 → 8, while the sliding scale truncates. Both are ported as canon has them; a test pins each.

### The anchoring invariant (acceptance criterion 1)

`SessionTransitions.test.ts` runs a property test over **200 deterministic pseudo-random pause/resume sequences** (seeded LCG, up to 24 operations each, 1–600 s between them), asserting at *every* step that elapsed equals the time independently tracked as "actually running", that elapsed never goes backwards, and that the anchor stays structurally consistent. A second block runs 60 more sequences with an app **kill inserted at a random point**, serialising the anchor through JSON and reviving it. A failure prints its seed and full operation trace.

### Threshold boundary (acceptance criterion 3's neighbour)

Canon's comparison is `elapsed < target * 0.3`, so on a 25-minute target the bar is 450 s and **exactly 450 s records** while 449.999 s does not. Both directions are asserted, as is the stopwatch case (always clears — canon `&&`-guards the threshold with `mode == .countdown`).

### Canon pin

Re-fetched `zheref/KroApple` `origin/main` before building: **`2c1ee45c09dc3e5fd5eed0e42709d377d640d1d2`** — identical to epic #1's pin. **No canon drift** since the epic was authored.

### Naming decisions

Three canon names already exist in the `@kro/core` barrel, exported by the legacy `/session` timer under `model/Session/` and imported today by `apps/web`. A second export of any of them is `TS2308`, so the **incoming** type moves. This is #7's precedent (`Perform.SessionFragment` → `PerformFragment`) applied by an explicit rule: **rename only where the barrel already holds the name; keep canon's spelling everywhere else.**

| Canon | Here | Colliding with |
|---|---|---|
| `SessionTimerMode` | `FocusTimerMode` | legacy `enum SessionTimerMode` |
| `SessionConfig` | `FocusSessionConfig` | legacy `class SessionConfig` (constructed in `apps/web/src/app/session/page.tsx`) |
| `SessionFragment` | `FocusSessionFragment` | legacy `interface SessionFragment` **and** `PerformFragment` |

`PersistedRunningSession`, `PersistedSessionEndeavor`, `PersistedSessionPhase`, `SessionSummary`, `SessionLaunchRecommendation` and `PointsFormula` collide with nothing and keep canon's spelling exactly. The adjective is the product's own — `docs/Features/Session.md` says "focus session" throughout. `__tests__/index.test.ts` pins the coexistence, so a future tidy-up that drops a `Focus` prefix fails there with a reason rather than in an app nobody ran.

### Unit vocabulary — settled (#8's issue comment)

**Seconds are the canonical `TimeInterval`, everywhere in this domain.** Every duration, threshold and fragment span is a `TimeIntervalSeconds` and matches the number Swift would encode.

`utils/durations.ts` — whose `seconds()` / `minutes()` return **milliseconds** under docstrings claiming milliseconds — is **not touched**: it is outside this issue's file lane, its behaviour is locked by the `useSession` / `useSessionTimer` suites in `apps/web`, and nothing in `domain/` imports it. #7 already introduced `TimeIntervalSeconds` with the unit spelled into the type name for exactly this reason. **Recommendation:** #21 rebuilds those two hooks on this domain and is the change that removes their last caller, so it can retire the millisecond helpers in the same PR that proves nothing still needs them.

### Canon divergences found

**1. Aborted attempts: recorded, or not? The two feature docs disagree.** `docs/Features/Performances.md` says a below-threshold finish and an abort produce **no performance record** ("Aborted → Not recorded"). `docs/Features/Session.md` says the opposite: *"Aborts and below-threshold Countdown finishes are recorded as aborted attempts."* The shipped code agrees with Session.md — `SessionSetupFeature` dispatches `.taskMarkedComplete(resolution: .aborted, …)` on both paths, reaching `PerformanceService.recordSessionPerformance`. Epic #1 makes **code the tie-breaker**, so this port follows the code: an aborted attempt **is** recorded, and earns zero under both formulas. This is not cosmetic — `empiricalDurationPerformances` filters aborted rows out of duration learning, which is precisely how canon keeps a half-hearted attempt visible in history while keeping it out of the recommendation. A "no record" reading would lose that. *(Recorded in `SessionThreshold.ts`'s module doc.)*

**2. `RewardCalculator.swift` contradicts itself on the same point.** Its `calculatePoints` doc comment reads `` `aborted`: 0% (deprecated - aborts should not create performance records) `` — while `SessionSetupFeature`, in the same repo, creates them. Flagging for KroApple; the *behaviour* (0 points) is unambiguous and is what is ported.

**3. The `Resolution` enum's own comments are misleading, and #7 inherited them.** `Endeavor.swift` documents `complete` as *"Task completed successfully"* and `finished` as *"Task finished early"*. Both `docs/Features/Performances.md` and `RewardCalculator.swift` use them the other way round: **`finished` = the task was marked done** (100%), **`complete` = the session ended with the task still open** (30%). Getting these backwards silently turns a 30% award into 100%, so it is worth naming. Our `packages/core/src/domain/endeavor/Perform.ts` carries the misleading Swift wording verbatim from #7. **It is outside this issue's file lane and is not touched here**; `RewardCalculator.ts`'s module doc states the correct reading, and its tests pin it. Suggest a one-comment fix in whichever child next touches `domain/endeavor`.

**4. `Open Space` is not a zero-duration preset.** Canon declares it `.init(title: "Open Space", mode: .stopwatch)` with no duration, so it inherits `SessionConfig.init`'s `3.hours` default — it carries a 3-hour `targetDuration`, not `0`. That is deliberate (the setup sheet can toggle back to countdown without losing a configured fallback, which `SessionLaunchRecommendation`'s own doc comment confirms), and it is ported as-is. Worth flagging because "stopwatch preset" reads like "no duration".

**5. `recordQuickComplete` writes `resolution: "complete"`,** while Performances.md's table classifies a quick complete as `finished` at 100%. Only the persistence layer is affected (#10's lane), not the formula — `calculateSlidingScalePoints` with `finished` + target 0 awards the full base, matching the doc. Flagged, not acted on.

### Behaviours ported as-is rather than "fixed"

Each of these is a place where a well-meaning repair would have silently made web and iOS disagree:

- **`remainingDuration` under stopwatch** returns `target − elapsed` (clamped) and canon's own comment says it is not meaningful there. Ported unchanged; `isRunningSessionCountdownFinished` is the mode-aware predicate a caller actually wants.
- **Duration bounds are applied floor → min → max**, so a `maximumDuration` below the `minimumDuration` wins.
- **Pausing a break loses the break phase.** `applySessionPaused` sets `.paused` unconditionally, so resuming reads `.paused` and returns to `.running`, not `.break`. Pinned by a test that says why.
- **`baseRewardFor` returns a flat 30** for every endeavor. `sessionPoints` now exists on our `Endeavor` (#7 ported it), but canon's branch that would read it is still commented out; honouring it here would make the two platforms award differently for the same endeavor.

### Deviations from the brief

- The brief summarised the reward rules as "aborted → NO record at all". As set out above, the two canon docs conflict and the code — the epic's tie-breaker — records an aborted attempt. Flagging explicitly since the brief said the doc wins: it does, but *which* doc had to be resolved, and Session.md plus the code agree against Performances.md.
- The brief described Open Space as a stopwatch preset with no duration; canon gives it the 3-hour init default (divergence 4).
- `SessionSummary` drops canon's `debugDescription` (locale formatting, which #7 excluded from this tier) and `asEKEvent` (EventKit has no web counterpart; epic #1 puts Apple EventKit hosts out of scope). `sessionSummaryStart` / `sessionSummaryEnd` expose the two values `asEKEvent` needed, so a calendar Service can be built on top without either import.
- `__tests__/barrel.test.ts` was renamed to `__tests__/index.test.ts`: the guard hook's barrel exemption greps file *content* for `class`, which this port's doc-comment table legitimately mentions. Giving the barrel a real test is the honest fix rather than rewording the doc — and the test earns its place, pinning the `TS2308` coexistence.

### Lane and hygiene

Touched **only** `packages/core/src/domain/session/**`, the `packages/core/src/index.ts` export block (anchored immediately after `domain/endeavor`, as the issue directs) and `packages/core/src/mocks.ts`. No new dependencies; **`pnpm-lock.yaml` is untouched**. `domain/endeavor`, `packages/app`, `apps/web`, root configs, `spec/`, `.claude/` and `docs/` are all unmodified. Nothing committed with `--no-verify`.

---

Bankai-Agent: ichigo

### CODE REVIEW

File: packages/core/src/domain/session/SessionTransitions.ts [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+    if (fragment !== undefined && fragment.end === null) {
+      target = index
+      break
+    }
+  }
+  if (target === -1) return fragments
+  return fragments.map((fragment, index) =>
+    index === target ? { start: fragment.start, end: endedAt } : fragment,
+  )
+}
```
[the remark this answers is not in the document]
Comment (zheref) [reply packages/core/src/domain/session/SessionTransitions.ts:43]: Fixed at fbf72c1 — transitions now close EVERY open fragment (withOpenFragmentsClosed). On every reachable state this is identical to canon single-close (activation only appends when nothing is open); on a corrupted anchor it self-heals instead of accruing elapsed forever. Two new tests pin the heal and the post-heal freeze; 720 core tests green.

**Comment by zheref** [comment]
**Copilot round dispositioned** (1 thread): defensive close-all on transitions, fixed at `fbf72c1` with self-heal tests. Orchestrator also corrected the swapped Resolution doc comments in domain/endeavor (`8b8c804` — cross-lane comment-only, flagged by this PR's audit). The five canon divergences are dispositioned: doc contradictions filed upstream on KroApple; the quick-complete resolution mismatch routed to #10.

**Fast-tracked** under the maintainer's express grant (mode → zheref/bankai-core#903): category (a) domain logic — merged by the orchestrating session.
